### ISSUE TITLE
Add interface for application register barriers inside the core

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#3798]
When the DR core performs certain "mangling" operations (application transformations required for core operation), it needs to read application register values.  This can conflict with drreg and clients in general who may be storing app values in memory temporarily.  This issue is asking for some kind of interface where the core can interact with drreg and clients to request a barrier where all values need to be in registers.

Today, for rseq (#2350), such a barrier is assumed to be in place at the start and end of blocks.  drreg obeys that today.  In the future we may want to support longer-range optimizations and trace optimizations, however.

This issue also covers documenting these implicit barriers.

**Comment by johnfxgalea** [comment]
> This issue is asking for some kind of interface where the core can interact with drreg and clients to request a barrier where all values need to be in registers.

Doesn't #4120 address this issue?

**Comment by derekbruening** [comment]
> > This issue is asking for some kind of interface where the core can interact with drreg and clients to request a barrier where all values need to be in registers.
> 
> Doesn't #4120 address this issue? or rather help in some way?

Yes, it helps.  I think what is left is how to handle the direction of the core calling into drreg.  Maybe the core provides an event or something that drreg (and a client doing other custom spilling) can register and the core can invoke to execute the barrier.  In drreg it would just run the new restore-all routine.
